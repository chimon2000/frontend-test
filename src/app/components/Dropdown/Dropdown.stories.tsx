import * as React from "react";
import { storiesOf } from "@storybook/react";
import { withInfo } from "@storybook/addon-info";
import { action } from "@storybook/addon-actions";
import { boolean, text } from "@storybook/addon-knobs";

import { Dropdown } from "./Dropdown";

const demoItems = ["sushi", "salad", "lemonade"];

const stories = storiesOf("Components", module);

stories.add(
  "Dropdown",
  withInfo({ inline: true })(() => (
    <Dropdown onUpdate={action("Menu Toggled")} open={boolean("open", false)}>
      {demoItems.map((item: string, idx: number) => (
        <div key={idx}>{item}</div>
      ))}
    </Dropdown>
  ))
);
