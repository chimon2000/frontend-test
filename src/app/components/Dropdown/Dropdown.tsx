import * as React from "react";

import Icon from "../Icon";
import { joinClassList } from "../../helpers";

export interface DropdownProps {
  /** The selection list. */
  children: React.ReactNode;
  /**
   * Whether to explicitly open or close the dropdown. Used for things like OnClickOutside.
   */
  open?: boolean;
  /**
   * Optionally provide a callback for open/closed trigger.
   */
  onUpdate?: (openStatus: boolean) => void;
  id?: string;
  className?: string;
}

export const Dropdown: React.FunctionComponent<DropdownProps> = ({
  open = false,
  onUpdate = () => {},
  className,
  id,
  children
}) => {
  const classes = [`dropdown`, `${className || ""}`];
  const elementRef = React.useRef(null);

  const [isOpen, toggleOpen] = React.useState(open);

  React.useEffect(() => toggleOpen(open), [open]);

  return (
    <div className={joinClassList(classes)} ref={elementRef}>
      <a
        className="dropdown__link"
        onClick={() => {
          toggleOpen(!isOpen);
          onUpdate(isOpen);
        }}
      >
        Click me
        <Icon className="dropdown__link-icon">
          {isOpen ? "keyboard_arrow_up" : "keyboard_arrow_down"}
        </Icon>
      </a>
      <ul
        className={`dropdown__menu ${isOpen ? "dropdown__menu--open" : ""}`}
        role="menu"
        data-open={isOpen}
        // style={{ display: isOpen ? "block" : "none" }}
      >
        {React.Children.map(children, (el: any, idx: number) => (
          <li key={idx}>{el}</li>
        ))}
      </ul>
    </div>
  );
};
