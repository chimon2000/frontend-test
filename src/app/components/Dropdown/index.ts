export { Dropdown as default, DropdownProps } from "./Dropdown";
