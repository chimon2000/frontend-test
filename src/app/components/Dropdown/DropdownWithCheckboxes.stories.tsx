import * as React from "react";
import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";
import { boolean, text } from "@storybook/addon-knobs";

const demoItems = ["sushi", "salad", "lemonade"];

const stories = storiesOf("Components", module);

import { Dropdown } from "./Dropdown";
import Checkbox, { ChangeEventObject } from "../Checkbox";

interface CheckboxList {
  [name: string]: boolean;
}

interface DemoProps {
  items: CheckboxList;
}
const items: CheckboxList = {
  sushi: true,
  salad: false,
  banana: false
};

export const DropdownWithCheckboxes: React.FunctionComponent = () => {
  const [changedItems, updateItems] = React.useState(items);

  const changeHandler = ({ name, value }: ChangeEventObject) => {
    updateItems({
      ...changedItems,
      [name]: value
    });
  };

  return (
    <div className="grid">
      <Dropdown className="col-8" open name="Click Me">
        {Object.keys(items).map((item: string, idx: number) => (
          <Checkbox
            key={idx}
            id={`item-${idx}`}
            label={item}
            name={item}
            checked={changedItems[item]}
            onChange={changeHandler}
          />
        ))}
      </Dropdown>
      <div>
        <h4>Items result:</h4>
        {JSON.stringify(changedItems)}
      </div>
    </div>
  );
};

stories.add("Dropdown With Checkboxes", () => <DropdownWithCheckboxes />);
