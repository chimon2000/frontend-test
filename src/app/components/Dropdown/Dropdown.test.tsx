import * as React from "react";
import { shallow } from "enzyme";

import { Dropdown } from "./Dropdown";

describe("<Dropdown />", () => {
  it("doesn't blow up", () => {
    const wrapper = shallow(
      <Dropdown>
        <div></div>
      </Dropdown>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("has isOpen flag default to false", () => {
    const wrapper = shallow(
      <Dropdown>
        <div></div>
      </Dropdown>
    );
    expect(wrapper.find('[data-open="false"]')).toBeDefined();
  });

  it("updates isOpen flag when clicked", () => {
    const wrapper = shallow(
      <Dropdown>
        <div></div>
      </Dropdown>
    );
    const link = wrapper.find(".dropdown__link");
    link.simulate("click");

    expect(wrapper.find('[data-open="true"]')).toBeDefined();
  });

  it("Calls onUpdate method when toggled closed/open", () => {
    let triggered = false;
    const updateMethod = () => {
      triggered = true;
    };
    const wrapper = shallow(
      <Dropdown onUpdate={updateMethod}>
        <div></div>
      </Dropdown>
    );
    const link = wrapper.find(".dropdown__link");
    link.simulate("click");
    expect(triggered).toEqual(true);
  });

  it("updates isOpen flag to true when explicitly passed", () => {
    const wrapper = shallow(
      <Dropdown open>
        <div></div>
      </Dropdown>
    );

    expect(wrapper.find('[data-open="true"]')).toBeDefined();
  });
});
