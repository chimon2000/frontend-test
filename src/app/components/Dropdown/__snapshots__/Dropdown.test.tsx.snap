// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> doesn't blow up 1`] = `
<div
  className="dropdown"
>
  <a
    className="dropdown__link"
    onClick={[Function]}
  >
    Click me
    <Component
      className="dropdown__link-icon"
    >
      keyboard_arrow_down
    </Component>
  </a>
  <ul
    className="dropdown__menu "
    data-open={false}
    role="menu"
  >
    <li
      key="0/.0"
    >
      <div />
    </li>
  </ul>
</div>
`;
